### PR TITLE
feat: Add maxWeight reduction to the MultiEigenStepperLoop

### PR DESCRIPTION
**Pull request by benjaminhuth**
This allows to choose between `mean` and `maxWeight` as reduction methods. A third method for the `mode` should be also added, but it does not yet work really well, and the `maxWeight` is also a bit of a "poor mans mode".

This PR makes this also configurable for the GSF in the examples.

Should be merged after #1627

### CODE REVIEW

File: Examples/Python/python/acts/examples/simulation.py [hunk 1]
```diff
@@ -352,6 +352,55 @@ def addParticleSelection(
     )
 
 
+def addParticleSelection(
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/simulation.py:355]: duplication from above?

File: Examples/Python/python/acts/examples/simulation.py [hunk 2]
```diff
@@ -378,8 +427,6 @@ def addFatras(
         the output folder for the Csv output, None triggers no output
     outputDirRoot : Path|str, path, None
         the output folder for the Root output, None triggers no output
-    rnd : RandomNumbers, None
-        random number generator
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/simulation.py:382]: is there a reason for removing this except that doc == arg?

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 3]
```diff
@@ -618,7 +618,7 @@ def addKalmanTracks(
         inputInitialTrackParameters="estimatedparameters",
         outputTrajectories="trajectories",
         directNavigation=directNavigation,
-        pickTrack=-1,
+        pickTrack=0,
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:621]: what's the difference between `0` and `-1`?

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp [hunk 4]
```diff
@@ -130,6 +130,12 @@ ActsExamples::ProcessCode ActsExamples::Geant4Simulation::execute(
   // Set the seed new per event, so that we get reproducible results
   G4Random::setTheSeed(m_cfg.randomNumbers->generateSeed(ctx));
 
+  // Set the seed new per event, so that we get reproducible results
+  G4Random::setTheSeed(m_cfg.randomNumbers->generateSeed(ctx));
+
+  // Set the seed new per event, so that we get reproducible results
+  G4Random::setTheSeed(m_cfg.randomNumbers->generateSeed(ctx));
+
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:138]: unintentional duplication?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp [hunk 5]
```diff
@@ -16,7 +16,14 @@ auto MultiEigenStepperLoop<E, R, A>::boundState(
   if (numberComponents(state) == 1) {
     return SingleStepper::boundState(state.components.front().state, surface,
                                      transportCov, freeToBoundCorrection);
-  } else {  // Do the combinatio
+  } else if (m_finalReductionMethod == FinalReductionMethod::eMaxWeight) {
+    auto cmpIt = std::max_element(
+        state.components.begin(), state.components.end(),
```
Comment (andiwand) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:21]: is it safe to assume `begin != end`?

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp (hunk 4 above)
Comment (benjaminhuth) [reply Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:138]: yeah that should not be here actually. reminder from a earlier PR

File: Examples/Python/python/acts/examples/reconstruction.py (hunk 3 above)
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/reconstruction.py:621]: this is a leftover

File: Examples/Python/python/acts/examples/simulation.py (hunk 2 above)
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/simulation.py:382]: hmm this seems to bie a leftover as well

File: Examples/Python/python/acts/examples/simulation.py (hunk 1 above)
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/simulation.py:355]: also a leftover

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp (hunk 5 above)
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:23]: I've added a assertion to check this

**Review by andiwand — APPROVED** [review]

**Review by andiwand — APPROVED** [review]

**Pull request merged**